src/main/scala/uk/gov/nationalarchives/signcookies/Lambda.scala

```diff
@@ -0,0 +1,143 @@
+import java.security.KeyFactory
+import java.security.spec.PKCS8EncodedKeySpec
+import java.time.temporal.ChronoUnit
+import java.util.{Base64, Date, UUID}
+import scala.concurrent.ExecutionContextExecutor
+import scala.io.Source
+
+class Lambda extends RequestStreamHandler {
+  val logger: Logger = Logger(this.getClass)
+  implicit val executionContext: ExecutionContextExecutor = scala.concurrent.ExecutionContext.global
+
+  private def decryptVariables(config: Config): Config = {
+    val encryptionContext = Map("LambdaFunctionName" -> config.functionName)
+    val kmsUtils = KMSUtils(kms(config.kmsEndpoint), encryptionContext)
+    config.copy(
+      privateKey = kmsUtils.decryptValue(config.privateKey),
+      authUrl = kmsUtils.decryptValue(config.authUrl),
+      frontendUrl = kmsUtils.decryptValue(config.frontendUrl),
+      environment = kmsUtils.decryptValue(config.environment),
+      uploadDomain = kmsUtils.decryptValue(config.uploadDomain),
+      keyPairId = kmsUtils.decryptValue(config.keyPairId)
+    )
+  }
+
+  private def validateToken(authUrl: String, token: String): IO[Token] = {
+    implicit val deployment: TdrKeycloakDeployment = TdrKeycloakDeployment(authUrl, "tdr", 60)
+    val e = KeycloakUtils().token(token)
+    IO.fromEither(e)
+  }
+
+  private def createResponse(cookies: CookiesForCustomPolicy, origin: String, config: Config): IO[String] = {
+    val originResponseHeaderValue = if (config.environment == "intg" && origin == "http://localhost:9000") {
+      origin
+    } else {
+      config.frontendUrl
+    }
+
+    val suffix = "Path=/; Secure; HttpOnly; SameSite=None"
+    val cookieResponse = List(
+      s"${cookies.getPolicy.getKey}=${cookies.getPolicy.getValue}; $suffix",
+      s"${cookies.getKeyPairId.getKey}=${cookies.getKeyPairId.getValue}; $suffix",
+      s"${cookies.getSignature.getKey}=${cookies.getSignature.getValue}; $suffix"
+    )
+    val headers = ResponseHeaders(originResponseHeaderValue, "true")
+    val multiValueHeaders = ResponseMultiValueHeaders(cookieResponse)
+    val response = LambdaResponse(200, headers.some, multiValueHeaders.some)
+    IO(response.asJson.printWith(Printer.noSpaces))
+  }
+
+  private def createCookies(userId: UUID, config: Config): IO[CookiesForCustomPolicy] = {
+    val decodedCert = Base64.getDecoder.decode(config.privateKey)
+    val keySpec = new PKCS8EncodedKeySpec(decodedCert)
+    val keyFactory = KeyFactory.getInstance("RSA")
+    val privateKey = keyFactory.generatePrivate(keySpec)
+    // Authorisation rule: only allow users to upload files to a folder corresponding to their user ID
+    val s3ObjectKey = s"$userId/*"
+    val protocol = Protocol.https
+    val distributionDomain = config.uploadDomain
+    val keyPairId = config.keyPairId
+    val activeFrom = Date.from(new Date().toInstant.minus(3, ChronoUnit.HOURS))
+    val expiresOn = Date.from(new Date().toInstant.plus(3, ChronoUnit.HOURS))
+    val ipRange = "0.0.0.0/0"
```

Comment:
Did you look into making this match the user's IP? It seems like a good security restriction.

Reply:
I didn't realise the sourceIp was sent in the request. I've added in the source IP to the case class and then add it to the policy. The test which is checking for the cookie policy now checks for this as well.